micro-rdk/src/esp32/nvs_storage.rs

```diff
@@ -126,6 +130,40 @@ impl NVSStorage {
     }
 }
 
+const BYTES_PER_ENTRY: usize = 32;
+
+impl StorageDiagnostic for NVSStorage {
+    type Error = NVSStorageError;
+    fn log_space_diagnostic(&self) -> Result<(), Self::Error> {
+        let mut stats: nvs_stats_t = Default::default();
+        esp!(unsafe { nvs_get_stats(self.partition_name.as_ptr(), &mut stats as *mut _) })?;
+
+        let used_entries = stats.used_entries;
+        let used_space = used_entries * BYTES_PER_ENTRY;
+        let total_space = stats.total_entries * BYTES_PER_ENTRY;
+
+        // From experimentation we have found that NVS requires 4000 bytes of
+        // unused space for reasons unknown. The percentage portion of the calculation (0.976)
+        // comes from the blob size restriction as stated in the ESP32 documentation
+        // on NVS
+        let total_usable_space = (0.976 * (total_space as f64)) - 4000.0;
+        let percentage_used = (used_space as f64) / total_usable_space;
+        log::info!(
+            "Credential storage stats: {:?} of {:?} bytes of NVS storage used",
+            used_space,
+            total_space
+        );
+        if percentage_used > 0.9 {
```

Comment:
If you used just `log!` could you compute a level, instead of having two logs?

Something like:

```
log!(if percentage_used > 90 { Level::Warn } else { Level::Info },
       "NVS stats {:?} bytes used of {:?} available");
```

Then you get a warning on high utilization and you don't need to say the "consider examination".